source/Missions/Services/Agents/Messages/AgentPublisher.cs

```diff
@@ -0,0 +1,132 @@
+﻿using Common.Logging;
+using Common.Messaging;
+using Missions.Services.Network;
+using Serilog;
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using TaleWorlds.MountAndBlade;
+
+namespace Missions.Services.Agents.Messages
+{
+    /// <summary>
+    /// Manages player controlled <see cref="AgentMovement"/>.
+    /// </summary>
+    public class AgentPublisher
+    {
+        private Agent _agent;
+
+        private Guid _agentId;
+
+        private AgentMovement _agentMovement;
+
+        private static readonly ILogger _logger = LogManager.GetLogger<AgentPublisher>();
+
+        private IMessageBroker _messageBroker;
+
+        private readonly CancellationTokenSource _agentPollingTaskCancellationTokenSource = new CancellationTokenSource();
+
+        private readonly Task _agentPollingTask;
+
+        private readonly int _packetUpdateRate;
+
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        public AgentPublisher(IMessageBroker messageBroker, int packetUpdateRate)
+        {
+            _agent = Agent.Main;
+            if (NetworkAgentRegistry.Instance.TryGetAgentId(_agent, out _agentId))
+            {
+                _logger.Warning($"Could not find Agent.Main id");
+                _agentId = Guid.NewGuid();
+
+                // TODO: after merge this must be changed to RegisterPlayerAgent
+                NetworkAgentRegistry.Instance.RegisterControlledAgent(_agentId, _agent);
```

Comment:
Use the DI agent registry
```suggestion
                _networkAgentRegistry.RegisterControlledAgent(_agentId, _agent);
```